pulsar-io/kafka/src/main/java/org/apache/pulsar/io/kafka/KafkaBytesSource.java

```diff
@@ -43,14 +49,64 @@
 
     @Override
     protected Properties beforeCreateConsumer(Properties props) {
-        props.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
-        props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+        props.putIfAbsent("schema.registry.url", "http://localhost:8081");
+        props.putIfAbsent(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class.getName());
+        props.putIfAbsent(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+
+        String currentValue = props.getProperty(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG);
+
+        // replace KafkaAvroDeserializer with our custom implementation
+        if (currentValue != null && currentValue.equals(KafkaAvroDeserializer.class.getName())) {
+            props.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, NoCopyKafkaAvroDeserializer.class.getName());
+        }
+
         log.info("Created kafka consumer config : {}", props);
         return props;
     }
 
     @Override
-    public byte[] extractValue(ConsumerRecord<String, byte[]> record) {
-        return record.value();
+    public Object extractValue(ConsumerRecord<String, Object> consumerRecord) {
+        Object value = consumerRecord.value();
+        if (value instanceof BytesWithAvroPulsarSchema) {
+            return ((BytesWithAvroPulsarSchema) value).getValue();
+        }
+        return value;
+    }
+
+    @Override
+    public org.apache.pulsar.client.api.Schema<?> extractSchema(ConsumerRecord<String, Object> consumerRecord) {
+        Object value = consumerRecord.value();
+        if (value instanceof BytesWithAvroPulsarSchema) {
+            return ((BytesWithAvroPulsarSchema) value).getPulsarSchema();
```

Comment:
PulsarAvroSchemaCache should be maintained at the connector level, not in the deserializer.

Reply:
unfortunately I don't have a way to pass the PulsarAvroSchemaCache instance to the deserializer (because we do not have control over the lifecycle).

Also we also need one instance and its lifecycle is tied to the lifecycle of the deserializer, so IMHO it can live internally to the deserializer.

Actually this deserializer produces `BytesWithAvroPulsarSchema` instances. I believe it is fine to leave all of the schema management stuff into it.

Reply:
The deserializer is constructed by reflection. There is a lot of unknowns on how Kafka would construct the deserializer. Hence I would suggest to keep the deserializer as simple as possible. What you need is just to get the schema id and the remaining content. See example at https://github.com/streamnative/pulsar-io-kafka/blob/master/src/main/java/io/streamnative/connectors/kafka/serde/KafkaSchemaAndBytesDeserializer.java

You don't need to pass the cache into the deserializer. You can implement the logic of extracting schema in `extractSchema` method. So the schema cache can be maintained in the connector, and then also you can get the schema information in the same pulsar thread rather than hiding the logic deeply into a deserializer where you don't know where the deserialize method will be run.

Reply:
done